### DISCUSSION TITLE
Camera path animation?

### DISCUSSION THREAD

**Discussion by clochardM33** [google/model-viewer#3293] [Q&A] [answered]
Is there any way to animate a camera position (implicitly as though it were a model animation)?
This would be to have a canned fly-by or similar.

Moving camera locations on a timer event (not that I've tried it) is a bit clunky for me.

Thanks,

**Comment by elalish** [comment]
glTF does suppport animated cameras, but we currently don't read them. Part of the reason is it's a bit unclear how to handle the situation where the glTF camera's aspect ratio doesn't match your element's. I would be open to a PR to add this functionality if you're interested.

**Reply by clochardM33** [reply]
Hi. That would be really good and yes I am interested.
I see what you mean about the ambiguity of the aspect ratio. In the one use case I have (see below) its really just the animation of the path and rotation that is important. To a first approximation the fov could be whatever it was when the user last touched it. It would of course be nicer to interpret it from the embedded info. 
As a thought if the camera object focal length is a compulsory field then is there something from that that can be used as a snaps-to-this-condition at the start of the path? (Else some fall back that reads as one line in the help documentation so it's straight forward).

Here is a bit of the sort of thing (but not exactly this). At 48s to 1m05s there is a fly through of the piece of kit I'm involved with.
https://vimeo.com/418407004
The kit has a lot of long thin aspect ratio components and being scientific there is value to be had making it look technical. The fly through the centre line is a classic marketing element for this sort of stuff.
I know the video is VR, but model-viewer has let me do a much lower cost version with an equivalently effective interaction in the context of customer training and education. (I think I'm thinking of the mine cart ride in Indiana Jones). Plus being able to watch as things woosh by is cool.

**Reply by elalish** [reply]
I think I'd be fine saying we only match the vertical FoV (since that's what our API specifies) to the focal length and allow the horizontal FoV to match the element's aspect ratio. I think it's important for clarity to use all of what the glTF camera specifies rather than picking and choosing. I think all of our camera API should be ignored while a glTF camera is selected. The other question would be, should the default orbit camera retain its prior state when you switch back to it, or should it take the state of the current glTF camera (say with radius = 0)? I think probably the former (you can always get the second behavior with a little JS). 

Glad you're liking MV! Yes, I'd definitely like to see more advanced uses like this; we've always envisioned them, but haven't seen many users try it out. Examples would be valuable.

**Reply by clochardM33** [reply]
That all makes perfect sense.
I would offer that the default orbit camera is a different thing to an animation and as such it is reasonable that they don't interact, which is your former point.
Some screenshots attached. I will work on sharing a more interactive look, but it's in public/private (corp) no mans land.
They show (in a very abstract way) a framework for training material; you've bought a thing, some annotation has brief techy notes, the MV is used as a prop when talking through how it works. 
The initial request would end up as running a camera path animation either from a click event or as one of the list animations to give a "worm's eye view" of the equipment. The kit has a concept of flow through it and being "part of the flow" is a value-add that you can't get either for real or in slides.
Which creates another FR: could you expose what is an what isn't a camera animation so they could be triggered differently?

![Screen3 - Desktop annotation active](https://user-images.githubusercontent.com/88084694/159757180-cafb8d75-36d3-4f47-b345-81fe9383445a.png)
![Screen3 - Mobile annotation active](https://user-images.githubusercontent.com/88084694/159757183-8827f09d-2ae5-468b-9994-429a7f35c16b.png)
![Screen1 - Desktop Simple Test Obect](https://user-images.githubusercontent.com/88084694/159757184-0d7f68ae-6e88-43c3-b693-56663d9c656a.png)
![Screen2 - Mobile Simple Test Obect](https://user-images.githubusercontent.com/88084694/159757186-b42157a1-47b4-463a-a8e4-7c79cba758f8.png)

**Reply by elalish** [reply]
Well, you can probably select the camera animation by name easily enough. Do you mean you want two different animations to run simultaneously?

**Reply by clochardM33** [reply]
I hadn't thought any of this through, but yes it would mean 1 camera and 1 geometry animation can run simultaneously. A bit like an animation with a camera orbit. Thanks.

**Reply by clochardM33** [reply]
Hi,
I've linked you to a private repository which has a more interactive version. Notes in the Readme

**Reply by elalish** [reply]
I'm not sure glTF makes an distinction between types of animation. You might be better off just exporting a single animation with both the geometry and camera animated in sync. Less chance for error that way too.

**Comment by clochardM33** [comment, marked as answer]
It's been a while, but going back to this one...
Forget importing camera paths for the time being. 
I can get almost the right outcome by keyframing cameraTarget (and associated), but what would be a sensible approach for handling the interpolationDecay value?

The non-sensible approach is to guess and set it bigger than the keyframe interval.

If I want a linear transition between "keyframes" and I know the keyframe timing what would is the relationship of these to interpolationDecay (assuming that there is one)?
By linear I mean no ease in or ease out of the motion or no ease out at least.
So from point A it moves to point B and gets up to speed pretty much straight away, through to B at that speed and then when going from B to C it appears to move at a constant rate with no slow down/speed up.
This is ignoring things like distances between points could be different which would affect the actual speed.
Thanks.

Actually I might be asking the wrong thing.
Maybe the question is can you just define the transition time between current camera position and a new camera position  and set it so there is only linear movement?

**Reply by elalish** [reply]
No, you can think of the decay like a spring & damper. If you really want it exactly linear, you should just calculate for every frame (do your own interpolation) and use `jumpToCameraGoal()` inside your RAF callback. If you want keyframes, then just select the next one significantly before you get to the current one, so the camera never gets too close to its goal. That will keep the acceleration fairly steady.

**Reply by clochardM33** [reply]
Got it. That is plenty to go with. Thanks.
